include/sys/context.h

```diff
@@ -11,7 +11,7 @@ typedef struct ctx ctx_t;
 typedef struct mcontext mcontext_t;
 
 /*! \brief Checks if saved context belongs to user space. */
-bool user_mode_p(ctx_t *ctx);
+bool user_mode_p(ctx_t *ctx) __no_instrument_kgprof;
```

Comment:
The problem appears in  `mips_exc_handler` cause we are calling `user_mode_p` and our instrumentation can enable interrupts
```
__no_instrument_kgprof void mips_exc_handler(ctx_t *ctx) {
  thread_t *td = thread_self();
  assert(cpu_intr_disabled());

  bool user_mode = user_mode_p(ctx);
```
it can be solved:
```
__no_instrument_kgprof void mips_exc_handler(ctx_t *ctx) {
  thread_t *td = thread_self();
  assert(cpu_intr_disabled());

  td->td_idnest++;
  bool user_mode = user_mode_p(ctx);
  td->td_idnest--;
```
but imo the first solution seems better. What do you think about it?

Reply:
I don't understand.
`user_mode_p` is a simply function that only checks if magic bits in context are present.
I understand that `__cyg_profile_func_enter` is called from that function but interrupts are disabled in `mips_exc_handler` 
`mips_exc_handler` calls `intr_root_handler` which requires disabled interrupts.
I don't see where `__cyg_profile_func_enter` enables and disables interrupts (if you enable interrupts in that function you also need to disable them, otherwise `intr_root_handler` fails).
Please elaborate.

Reply:
Ok, in `mips_exc_handler` the value of `td->td_idnest` is 0 (counter how many times we have disabled interrupts), but the interrupts are still disabled and when we are disabling interrupts in our instrumentation we increase the counter and while leaving we decrease it and when we hit 0 we will enable interrupts. It is a similar problem like in `intr_root_handler`

```
__no_instrument_kgprof void intr_root_handler(ctx_t *ctx) {
  thread_t *td = thread_self();

  assert(cpu_intr_disabled());
  assert(td->td_idnest == 0);

  /* Increment interrupt disable nesting counter in order to prevent filter
   * routines to accidentaly enable interrupts. */
  td->td_idnest++;
```

Reply:
I added an assertion about the `td->td_inest` value. If you take a look at `intr_enable` and `int_disable` definition (which are used in the instrumentation) it should be clear why we do this.

Reply:
> I don't see where __cyg_profile_func_enter enables and disables interrupts (if you enable interrupts in that function you also need to disable them, otherwise intr_root_handler fails).

When we take a spinlock and release it we disable and enable interrupts.
And `intr_root_handler` does not fail cause we increase `td->td_inest` value.

Reply:
Okay, I see. It should be documented.